eth/downloader/downloader.go

```diff
@@ -549,19 +587,30 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td, ttd *
 		// the ancientLimit through that. Otherwise calculate the ancient limit through
 		// the advertised height of the remote peer. This most is mostly a fallback for
 		// legacy networks, but should eventually be droppped. TODO(karalabe).
-
-		// Legacy sync, use the best announcement we have from the remote peer.
-		// TODO(karalabe): Drop this pathway.
-		if remoteHeight > fullMaxForkAncestry+1 {
-			d.ancientLimit = remoteHeight - fullMaxForkAncestry - 1
+		if beaconMode {
+			// Beacon sync, use the latest finalized block as the ancient limit
+			// or a reasonable height if no finalized block is yet announced.
+			if final != nil {
+				d.ancientLimit = final.Number.Uint64()
+			} else if height > fullMaxForkAncestry+1 {
+				d.ancientLimit = height - fullMaxForkAncestry - 1
+			} else {
+				d.ancientLimit = 0
+			}
 		} else {
-			d.ancientLimit = 0
+			// Legacy sync, use the best announcement we have from the remote peer.
+			// TODO(karalabe): Drop this pathway.
+			if height > fullMaxForkAncestry+1 {
+				d.ancientLimit = height - fullMaxForkAncestry - 1
+			} else {
+				d.ancientLimit = 0
+			}
 		}
 		frozen, _ := d.stateDB.Ancients() // Ignore the error here since light client can also hit here.
-		itemAmountInAncient, _ := d.stateDB.ItemAmountInAncient()
+
 		// If a part of blockchain data has already been written into active store,
 		// disable the ancient style insertion explicitly.
-		if origin >= frozen && itemAmountInAncient != 0 {
```

Comment:
even frozen!=0
item amount in ancient may be zero
so keep use itemAmountInAncient?

Reply:
revert, eth/downloader change will be solved in next PR.

Reply:
revert, eth/downloader change will be solved in next PR.